feat(floodsub): make use of `prost-codec`

nloadholtes opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->
This patch addresses #2500 for the `libp2p-floodsub` crate.

For this PR the existing code was upgraded to use `Framed` with the `prost_codec::Codec` as the standard codec for handling the RPC message format serialization/deserialization.

## Notes

<!-- Any notes or remarks you'd like to make about the PR. -->

  - There are no pre-existing tests, so it is difficult to check if this is working correct. There were no errors observed locally.
  - The protocol name and message size were extracted into constants.
  - I'm still continuing my rust journey so suggestions/comments are welcomed.

## Links to any relevant issues

<!-- Reference any related issues.-->
#2500 

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

nloadholtes committed:
Initial update to floodsub

Extracing protocol name string, updating outbound to use prost_codec

Seems to compile ok, kinda makes sense

Cleanup

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -135,13 +149,22 @@ where
+            framed
+                .send(bytes)
+                .await
+                .map_err(|e| io::Error::new(std::io::ErrorKind::ConnectionRefused, e))?;
```
What type is `e` here?

nloadholtes replied:
Great question, and I'm not 100% sure. Its some kind of socket error, and I was wanting to use as general of a an error where as I could get away with. But I couldn't figure out how to get the compiler to accept a more "general" thing. VSCode kept offering suggestions of ConnectionRefused and other similar/likely errors (peer disconnected, etc.). I picked ConnectionRefused (and ConnectionReset below) knowing that we might be talking about them. ;)

I did try refactoring the `FloodsubDecodeError` to see if that could be expanded into more of a general "FloodsubError", but it really started to feel like scope creep on this ticket, at least in my opinion. I'd really like to do that approach (just feels more architecturally "correct" IMO), but I also want to keep this PR focused on the problem at hand.

I'm open to suggestions on all of this. (I also want to put some tests here, but first I need to crawl, then walk, then run :laughing: )

thomaseizinger replied:
I'd suggest we:

1. Create a type called `EncodeError` analogous to `DecodeError`.
2. Set `type Error = EncodeError` on line 149.
3. Remove these `map_err` calls.

thomaseizinger reviewed (commented):
Looks good overall! Just one question, I am not quite happy with the error mapping there.

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -135,13 +149,22 @@ where
-
+            let mut framed = Framed::new(
+                socket,
+                prost_codec::Codec::<std::vec::Vec<u8>>::new(MAX_MESSAGE_LEN_BYTES),
```
Feeding bytes into this somewhat defeats the purpose of using `Framed` and `Codec`.

Instead, you'd want to change `self.into_bytes` to return you the `rpc_proto::Rpc` type and feed that into `framed.send`.

thomaseizinger replied:
You can then also remove this type hint, the compiler will figure it out!

```suggestion
                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
```

nloadholtes replied:
Hmmm, I kept getting errors about the type after I removed it. I looked elsewhere in the code and [saw in indentify ](https://github.com/libp2p/rust-libp2p/blob/c1e68b75c0729102ee43e2ec3cedc643b4d1f324/protocols/identify/src/protocol.rs#L181-L190)there was a section of code using a similar pattern of adding a type hint when created its `FramedWrite`.

Also in that code is the calls to the `send()` and `write()` functions, but for some reason that code does not generate warning about unsued returns like my code does :thinking: I did see that `protocols/gossipsub/src/config.rs` uses this pattern, but there it makes sense: the code is just setting fields and doesn't care about the result. For my code, I don't know why the compiler suddenly cares about the return of those 2 calls.

thomaseizinger replied:
Right, my bad! It is needed :)

thomaseizinger reviewed (commented):
Getting there! A few more comments :)

Commit:
build rpc, new error type, simplification of code

Commit:
Readme update, removing unused dependency, changing upgrade_outbound to have type_hint and clearing out warnings about unused returns

Commit:
Merge branch 'master' into 2500-floodsub

Commit:
CHANGELOG fixes

thomaseizinger commented on protocols/floodsub/CHANGELOG.md:
```diff
@@ -2,6 +2,12 @@
 
+- Update to `prost-codec` `v0.3.0`.
+
+- Refactoring Floodsub to use common protobuf Codec. See [PR 3224].
```
```suggestion
- Read and write protocols messages via `prost-codec`. See [PR 3224].
```

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -132,25 +149,26 @@ where
+                prost_codec::Codec::<rpc_proto::Rpc>::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            let _ = framed.send(self.into_rpc()).await;
+            let _ = framed.close().await;
```
```suggestion
            framed.send(self.into_rpc()).await?;
            framed.close().await?;
```

We should not ignore those errors :)

nloadholtes replied:
Ahh! This just solved a mystery I ran into with the error structs/enums. I am refactoring some stuff :smile:

nloadholtes replied:
My confusion: (In case I look at this PR again in the future)  the error message that adding the `?` generated was not clear to me. What it was asking was for the `Error` type to handle the output of the `?` and what I had there originally (just the struct) did not handle the prost_code:Error error types. So of course that error message now makes sense. The `type Error` needs be aware of any errors that could come out of the impl.

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -56,10 +61,18 @@ where
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
```
```suggestion
            let mut framed = Frame::new(
                socket,
                prost_codec::Codec::<rpc_proto::Rpc>::new(MAX_MESSAGE_LEN_BYTES),
            );
```

This should be sufficient.

thomaseizinger reviewed (changes requested):
Looks good, almost there :)

nloadholtes committed:
Update protocols/floodsub/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

Commit:
Streamline error handling, refactoring internal names, ordering fields in enum

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -93,21 +106,24 @@ where
 #[derive(thiserror::Error, Debug)]
 #[error(transparent)]
-pub struct DecodeError(prost::DecodeError);
+pub struct EncodeDecodeError(prost_codec::Error);
```
```suggestion
pub struct CodecError(prost_codec::Error);
```

nloadholtes replied:
haha! I put that name down first, but changed my mind about 5 times before settling. I agree and will change it.

thomaseizinger reviewed (approved):
Just one naming suggestion, otherwise LGTM!

thomaseizinger commented on protocols/floodsub/src/protocol.rs:
```diff
@@ -93,21 +106,24 @@ where
+pub enum FloodsubError {
+    /// Error from prost_codec::Codec
+    #[error(transparent)]
+    Codec(#[from] prost_codec::Error),
```
This one should not exist! We want to hide `prost_codec::Error` from the public API of this crate!

thomaseizinger replied:
That is the reason for `CodecError` existing :)

nloadholtes replied:
umm, I just got side tracked on this. While trying to fix it I caused an error ` the compiler unexpectedly panicked. this is a bug.` I'm going to go report that and then pick this up again tomorrow. Thanks for all your help on this @thomaseizinger!

thomaseizinger replied:
> umm, I just got side tracked on this. While trying to fix it I caused an error ` the compiler unexpectedly panicked. this is a bug.` I'm going to go report that and then pick this up again tomorrow. Thanks for all your help on this @thomaseizinger!

`cargo clean` usually fixes these. I hit them too occasionally, often in combination with proc-macros.

nloadholtes replied:
That did the trick! I cleaned up the commit message and pushed what I think is the corrected code.

Commit:
Better name for the Error struct

Commit:
Removed prost_codec from FloodsubError

This prevents exposing prost_codec publicly (changing the API). Changed
CodecError to be from prost_codec so that the upgrade_outbound can
use it as its Error type.

thomaseizinger reviewed (dismissed):
Thanks!

mergify[bot] committed:
Merge branch 'master' into 2500-floodsub

thomaseizinger committed:
Merge branch 'master' into 2500-floodsub

thomaseizinger committed:
Merge branch 'master' into 2500-floodsub

mergify[bot] committed:
Merge branch 'master' into 2500-floodsub

mxinden reviewed (approved):
:rocket: Thanks @nloadholtes!
